lib/db.js

```diff
@@ -22,10 +23,91 @@ class DB {
                 statement.finalize();
             }
         });
+        /* Process the array of storage groups declared in the config */
+        this._storageGroups = this._buildStorageGroups(this.conf.storage_groups);
+        /* The cache holding the already-resolved domain-to-group mappings */
+        this._storageGroupsCache = new Map();
+    }
+
+    /**
+     * Process the storage group configuration.
+     * @param {Array} groups the array of group objects to read, each must contain
+     *                at least the name and domains keys
+     * @return {Array} Array of storage group objects
+     */
+    _buildStorageGroups(groups) {
+        const storageGroups = [];
+        if (!Array.isArray(groups)) {
+            return storageGroups;
+        }
+        groups.forEach((group) => {
+            const grp = extend(true, {}, group);
+            if (!Array.isArray(grp.domains)) {
+                grp.domains = [grp.domains];
+            }
+            grp.domains = grp.domains.map((domain) => {
+                if (/^\/.*\/$/.test(domain)) {
+                    return new RegExp(domain.slice(1, -1));
+                }
+                return domain;
+            });
+            storageGroups.push(grp);
+        });
+        return storageGroups;
```

Comment:
I wonder if both here and in mod-table-cass we should ensure that a catch-all group is present, since we rely on this fact throughout the code.

Reply:
M.... We actually don't have a catch-all group explicitly in, for example, RESTBase in production. Supplying a default `catch-all` group is tricky because what exactly are you going to name it? Given that we no longer auto-create the actual keyspaces, I believe that the right way would be to fail-fast and through if the storage group was not found on line 87. That would make the PR incompatible with old versions though. 

Before we didn't support the storage groups, so we implicitly used `domain -> domain` mappings for "storage groups" and that's what it is currently defaulting to on line 87. I think it's a decent-enough compromise between backwards compatibility and perfectionism

Reply:
We do have [the catch-all group in prod](https://github.com/wikimedia/mediawiki-services-restbase-deploy/blob/4b34370ea3f80ac0995333b756897e158d4f5a3d/scap/templates/config.yaml.j2#L87). I would personally prefer having a default or failing. Neither of these options would be incompatible with the previous status quo

Reply:
Oh.. indeed. Failfast it is!